Support fish completion

kri5 opened the pull request:
**What this PR does / why we need it**:

This PR delegates support for completion entirely to cobra.

This allows to support more shells without the need to maintain specific code on helm side.

closes #8109

kri5 committed:
Adds completion for fish shell

Signed-off-by: Christophe Courtaut <christophe.courtaut@gmail.com>

kri5 commented:
Sorry for the noise on this PR.

I first tried to change the name of the binary directly when the command is created, but it leads to tests not passing because the name is also used for usage, and tests hardcode "helm" binary name.

So i went for a method were i override "just-in-time" the name of the command, in order to generate completion with cobra only and get rid of any "completion" code specific to a shell in helm.

Then i added the support for the shells cobra already supports.

Feel free to comment

marckhouzam commented on cmd/helm/completion.go:
```diff
@@ -74,173 +74,23 @@ func runCompletion(out io.Writer, cmd *cobra.Command, args []string) error {
-}
-__helm_compopt() {
-	true # don't do anything. Not supported by bashcompinit in zsh
+	return cmd.GenZshCompletion(out)
```
Cobra 1.0's support for Zsh is a lesser solution than what helm currently provides.
The biggest issue is that it does not support custom completions, so
`helm status <TAB>` would no longer work, for example.

We first need https://github.com/spf13/cobra/pull/1070 to be merged and released before we can start using Cobra's Zsh completion.

kri5 replied:
Agreed

marckhouzam commented on cmd/helm/completion.go:
```diff
@@ -74,173 +74,23 @@ func runCompletion(out io.Writer, cmd *cobra.Command, args []string) error {
-	out.Write([]byte(zshTail))
-	return nil
+func runCompletionFish(out io.Writer, cmd *cobra.Command) error {
+	return cmd.GenFishCompletion(out, true)
```
What do you think of making the second parameter configurable to helm's users?
I suggest adding a `--no-descriptions` flag to the `completion` command.
You can get inspiration from the work started here: https://github.com/helm/helm/pull/7690

kri5 replied:
Will try to do that, i was not aware of the work you previously done on this.

marckhouzam replied:
No problem.  Contributions are always welcomed and help spread the knowledge.

kri5 replied:
should be fixed in c6849c4014c3524bca4b2dd6698594767fa8c386

marckhouzam commented on cmd/helm/completion.go:
```diff
@@ -52,7 +52,7 @@ func newCompletionCmd(out io.Writer) *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "completion SHELL",
-		Short: "generate autocompletions script for the specified shell (bash or zsh)",
+		Short: "generate autocompletions script for the specified shell (bash, zsh or fish)",
```
I think the long description also needs to include 'fish'

kri5 replied:
Agreed, tottaly missed that one

kri5 replied:
76c63b819325c72bfb4f6f7460ff52a83bfb9c10 should fix it

marckhouzam commented on cmd/helm/completion.go:
```diff
@@ -74,173 +74,23 @@ func runCompletion(out io.Writer, cmd *cobra.Command, args []string) error {
+	// both helm2 and helm3), we overwrite the new binary name to the command
+	// in order to generate proper completion
+	rootCmd := cmd.Root()
+	rootCmd.Use = filepath.Base(os.Args[0])
```
I'm a little concerned about unknown side-effects of doing this.
But it may be ok.  I'm just leaving a note to keep thinking about it.

kri5 replied:
We might want to create a deep copy of the command?
I haven't investigated much as if the command was still used afterwards.

marckhouzam replied:
Simpler than a deep-copy could be to
- set Use field
- generate completion script
- reset Use field to old value.

But all this feels hacky to me.

kri5 replied:
Seems hacky to me too, but i think that's a job for cobra, it definitly should be configurable.
Meanwhile we might want something like that.

I changed the way it was done in order to save/restore the Use field in fe31b1db37f1f1ee7452234a852aa286e2aea581

marckhouzam reviewed (changes requested):
I'm really excited about this contribution!  It is great to see others getting involved with Helm completion code.

Note that for the Fish completion code to also work with Cobra 1.0, we need PR #7906 to get merged.

marckhouzam commented:
I just remembered that as it stands, custom completions (`helm status <TAB>`) will not work for Fish.  We have to move all our custom completion code to Cobra 1.0.0.

I have a patch ready for that but it first needs #7906 to get merged.

kri5 committed:
Adds a no-descriptions flags for completions

This allow to choose if descriptions in completions
are generated for shells supporting it.

Signed-off-by: Christophe Courtaut <christophe.courtaut@gmail.com>

marckhouzam commented:
> I just remembered that as it stands, custom completions (`helm status <TAB>`) will not work for Fish. We have to move all our custom completion code to Cobra 1.0.0.
> 
> I have a patch ready for that but it first needs #7906 to get merged.

#7906 was merged 🎉 and I have posted the follow-up PR #8131 which is required to allow Fish completions to have custom completions.

marckhouzam commented:
@kri5, if you want to rework this PR to only focus on Fish completion (no changes to Bash or Zsh), we could bring it in right away, even before #8131. That would at least provide fish completion for commands and flags.

marckhouzam commented:
Hi @kri5 .  PR #8131 which moved all dynamic completions to Cobra's new framework has been merged.  This means we can get full Fish completion support very easily.

Will you have time to adapt your PR to limit it to Fish completion?  If not, I can take it over if it helps.  Thanks.

kri5 commented:
Sorry for the late response, there should be only fish completion related changes.
Feel free to comment :)

marckhouzam commented on cmd/helm/completion.go:
```diff
@@ -244,3 +247,7 @@ __helm_bash_source <(__helm_convert_bash_to_zsh)
 }
+
+func runCompletionFish(out io.Writer, cmd *cobra.Command) error {
+	return cmd.GenFishCompletion(out, !completionNoDesc)
```
Actually, this does not work.  You need to call `cmd.Root().GenFishCompletion(...)`.

marckhouzam reviewed (changes requested):
Almost there.

marckhouzam commented on cmd/helm/completion.go:
```diff
@@ -52,13 +54,14 @@ func newCompletionCmd(out io.Writer) *cobra.Command {
 		},
 		ValidArgs: shells,
 	}
+	cmd.Flags().BoolVar(&completionNoDesc, "no-descriptions", false, "disable completion descriptions for shells that support it")
```
It was bothering me that this flag would apply to all shells instead of just the fish shell.
I've opened #8314 to allow to improve on this.  Let's see what the other maintainers think of this.

kri5 replied:
Agreed! :+1:

marckhouzam commented:
Because we're waiting for #8134 let's target the 3.4 release for this improvement.

marckhouzam commented:
@kri5 could you rebase this PR to adapt to #8314 which has just been merged? You will then be able to add the `--no-descrptions` flag to the `fish` command only.

marckhouzam commented:
Hi @kri5. Just checking to see if you will have some time to work on the fish completion rebase.  There's no real urgency since it's targeted for the 3.4 release but I just wanted to ask.  Thanks.

marckhouzam commented:
Hi @kri5. With 3.3 released, work has begun for helm 3.4.  Will you be able to rebase this PR so we can provide fish completion?

If you are unable to, I can take it over if you prefer?

marckhouzam commented:
I have pushed a new PR, #8664, to replace this one, to make sure it makes it on time for the 3.4 release, with some soak time.

Thank you @kri5 for getting the ball rolling on this change.  Any future contribution is welcomed.
